design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,156 @@
+The following UDFs will be added:
+
+* `to_bytes(string, inputEncoding, outputEncoding)` - this will convert a STRING value in the specified encoding format to a BYTES in the specified encoding format.
+The allowed encoders are the same as the ones allowed in the existing `encode` function.
+* `decode(bytes, inputEncoding, outputEncoding)` - this will convert a BYTES value in the specified encoding format to a STRING in the specified encoding format.
+
+We will also update some of the existing STRING functions to accept BYTES as a parameter. In general, if a function works on ASCII characters for a STRING parameter,
+then it will work on bytes for a BYTES parameter.
+
+* `len(bytes)` - This will return the length of the stored ByteArray.
+* `concat(bytes...)` - Concatenate an arbitrary number of byte fields
+* `r/lpad(bytes, target_length, padding_bytes)` - pads input BYTES beginning from the left/right with the specified padding BYTES until the target length is reached.
+* `replace(bytes, old_bytes, new_bytes)` - returns the given BYTES value with all occurrences of `old_bytes` with `new_bytes`
+* `split(bytes, delimiter)` - splits a BYTES value into an array of BYTES based on a delimiter
+* `splittomap(bytes, entryDelimiter, kvDelimiter)` - splits a BYTES value into key-value pairs based on a delimiter and creates a MAP from them
+* `substring(bytes, to, from)` - returns the section of the BYTES from the byte at position `to` to `from`
+
+## Design
+### Serialization/Deserialization
+
+BYTES will be handled by [`java.nio.ByteBuffer`](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html) within ksqlDB.
+The underlying Kafka Connect type is the primitive `bytes` type. 
+
+#### Avro
+
+`bytes` is a primitive Avro type. When converting to/from Connect data, the Avro converter ksqlDB
+uses converts byte arrays to ByteBuffer.
+
+#### Protobuf
+
+`bytes` is a primitive Protobuf type. The maximum number of bytes in a byte array is 2<sup>32</sup>.
+When converting to/from Connect data, the Avro converter ksqlDB uses converts byte arrays to ByteBuffer.
+
+#### JSON/Delimited
+
+Byte arrays will be stored in JSON and CSV files as [Base64 MIME](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html#mime) encoded binary values.
+This is because ksqlDB and Schema Registry both use Jackson to serialize and deserialize JSON,
+and Jackson serializes binaries to Base64 strings.
+
+The ksqlDB JSON and delimited deserializers will be updated to convert Base64 strings to ByteBuffer.
+
+### Casting
+
+Casting BYTES to STRING will convert the BYTES value to a STRING value encoded by the format specified
+by `ksql.bytes.format`. Casting STRING to BYTES will convert the STRING to a BYTES value decoded by the
+format specified by `ksql.bytes.format`.
+
+Some other alternatives are:
+* Use UTF-8 for all casts, and throw if it fails (BigQuery does this)
+* Not support casting. This would make BYTES the only data type that cannot be cast to STRING.
```

Comment:
I vote for not supporting casting. Instead, users can use conversion functions with the specified format. Easier than handling configs.

Reply:
That sounds good to me.

Reply:
Updated KLIP to remove support for casting